src/Page/Form.purs

```diff
@@ -47,18 +54,105 @@ component =
   initialState :: State
   initialState =
     { form: Left "Loading..."
+    , values: mempty
     }
 
   eval :: Query ~> H.ComponentDSL State Query Message m
   eval (Initialize a) = a <$ do
+    let values = keysPage testPage
+    H.modify_ _ { values = values }
     void $ bitraverse_
       (\e -> H.modify _ { form = Left e })
       (\f -> eval $ EvalForm f a)
       (Right testPage)
 
-  eval (EvalForm form a) = a <$ do
-    H.modify_ _ { form = pure form }
-    pure unit
+  eval (EvalForm expr a) = a <$ do
+    let evaled' = evalPage get expr
+        get key = Data.Map.lookup key values
+        values = keysPage expr
+    H.modify_ _ { values = values }
+    case evaled' of
+      Left _ -> pure unit
+      Right evaled -> H.modify_ _ { form = pure { expr, evaled } }
+
+  eval (UpdateKey key val a) = do
+    form <- H.gets _.form
```

Comment:
This is a tiny, mostly cosmetic thing, but I like to write these as
```purescript
{ form } <- H.get
```
I think it's a little cleaner, and allows you to match multiple values more efficiently if you needed to do something like
```purescript
{ form, values } <- H.get
```